opensrp-anc/src/main/java/org/smartregister/anc/library/widget/ANCSpinnerFactory.java

```diff
@@ -0,0 +1,192 @@
+
+    public static final HashSet<String> locationSpinners = new HashSet<String>() {
+        {
+            add(PROVINCE);
+            add(DISTRICT);
+            add(SUB_DISTRICT);
+            add(FACILITY);
+            add(VILLAGE);
+        }
+    };
+
+    public static final Map<String, List<String>> descendants = new HashMap<String, List<String>>() {
+        {
+            put(PROVINCE, Arrays.asList(DISTRICT, SUB_DISTRICT, FACILITY, VILLAGE));
+            put(DISTRICT, Arrays.asList(SUB_DISTRICT, FACILITY, VILLAGE));
+            put(SUB_DISTRICT, Arrays.asList(FACILITY, VILLAGE));
+            put(FACILITY, Arrays.asList(VILLAGE));
+            put(VILLAGE, null);
+        }
+    };
+
+    private final Map<String, String> parents = new HashMap<String, String>() {
+        {
+            put(DISTRICT, PROVINCE);
+            put(SUB_DISTRICT, DISTRICT);
+            put(FACILITY, SUB_DISTRICT);
+            put(VILLAGE, FACILITY);
+        }
+    };
+    private JsonFormFragment formFragment;
+    private JsonFormActivity jsonFormView;
+
+    @Override
+    public void genericWidgetLayoutHookback(View view, JSONObject jsonObject, JsonFormFragment formFragment) {
+        super.genericWidgetLayoutHookback(view, jsonObject, formFragment);
+        View materialView = ((RelativeLayout) view).getChildAt(0);
+        materialView.setOnTouchListener((v, event) -> {
+            materialView.setTag(R.id.is_human_action, true);
+            return false;
+        });
+    }
+
+    @Override
+    public List<View> getViewsFromJson(String stepName, Context context, JsonFormFragment formFragment,
+                                       JSONObject jsonObject, CommonListener listener, boolean popup) throws Exception {
+        this.formFragment = formFragment;
+        jsonFormView = (JsonFormActivity) formFragment.getActivity();
+
+        try {
+            if (jsonObject.has(SUB_TYPE)
+                    && jsonObject.getString(SUB_TYPE)
+                    .equalsIgnoreCase(LOCATION_SUB_TYPE)
+                    && jsonObject.has(ConstantsUtils.JsonFormKeyUtils.OPTIONS)
+                    && jsonObject.getJSONArray(ConstantsUtils.JsonFormKeyUtils.OPTIONS).length() <= 0) {
+
+                String stepTitle = formFragment.getStep(STEP1).get(STEP_TITLE).toString();
+                if (stepTitle.equals(ConstantsUtils.EventTypeUtils.REGISTRATION)
+                        && StringUtils.endsWithIgnoreCase(jsonObject.getString(KEY), "province")) {
+                    populateProvince(jsonObject);
+                } else {
+                    populateDescendants(jsonObject);
+                }
+            }
+        } catch (JSONException e) {
+            Timber.e(e);
+        }
+
+        return super.getViewsFromJson(stepName, context, formFragment, jsonObject, listener, popup);
+    }
+
+    protected void populateProvince(JSONObject jsonObject) {
+        List<LocationTag> tags = Utils.getLocationTagsByTagName("Country");
```

Comment:
Move the string literals to the constants file

Reply:
Done